ARROW-8413: [C++][Parquet] Refactor Generating validity bitmap for values column

emkornfield opened the pull request:
This change does the following:
* Vectorizes computation for validity bitmaps with no repeated parents for all little-endian architectures 
* Vectorizes computation for validity bitmaps with repeated parents for AVX2 + architectures (really it requires BMI2)
* Exposes some building blocks to do vectorized computation for all validity bitmaps in for parquet level data (more need to be handled for generating appropriate bitmaps and offset data for lists).  These will be added level_conversions.h in a future PR.
* Replaces loops over bitmaps with SetBitsTo
* Leaves a fallback for non BMI2/little-endian capable machines.

With AVX2 enabled this seems to improve benchmarks by 20% for nullable columns (BM_Read..) on my box. I didn't see any impacts in other benchmarkmarks.

See checklist for what is remaining.  If possible i'd like to get early feedback on naming and my approach to checking for BMI2.

Still needed:
- [x] Still adding more focused unit tests for what I've added.
- [x] Move the changes to Bitmap::ToString to there own PR.

kiszk commented on cpp/src/parquet/CMakeLists.txt:
```diff
@@ -336,6 +336,7 @@ set_source_files_properties(public_api_test.cc
 add_parquet_test(reader_test
                  SOURCES
                  column_reader_test.cc
+		 level_conversion_test.cc
```
nit: indentation

emkornfield replied:
done.

kiszk commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,141 @@ using arrow::internal::checked_cast;
+    uint8_t* valid_bits, int64_t valid_bits_offset) {
+  if (max_repetition_level > 0) {
+#if ARROW_LITTLE_ENDIAN
+// we use AVx2 as a proxy for BMI2 instruction set, since there doesn't seem to be a clean
+// way o detect that latter for MSVC.
+#if defined(ARROW_HAVE_BMI2)
+    // We need BIM2 instruction which is AVX2 should imply.
```
nit: AVx2 -> AVX2, o detect -> to detect, BIM2 -> BMI2

emkornfield replied:
Remove the comment as it was stale.

fsaintjacques commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -43,13 +43,18 @@
 #define ARROW_BYTE_SWAP64 __builtin_bswap64
 #define ARROW_BYTE_SWAP32 __builtin_bswap32
+#define ARROW_POPCOUNT64 __builtin_popcountll
+#define ARROW_POPCOUNT32 __builtin_popcount
```
TIL, C++20 has std::popcount.

emkornfield replied:
Yep, so we'll be able to use it by 2030 hopefully? :)

pitrou replied:
I'd say 2027.

emkornfield commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -40,12 +40,13 @@ if(ARROW_CPU_FLAG STREQUAL "x86")
   else()
     set(ARROW_SSE4_2_FLAG "-msse4.2")
-    set(ARROW_AVX2_FLAG "-mavx2")
+    set(ARROW_AVX2_FLAG "-march=core-avx2")
```
hopefully this is acceptable change.  I'm open to other suggestions on how to enable bmi2 (not for some reason using "-mbmi2 -mavx2" didn't seem to work for me, but I can try again.

pitrou replied:
It would be nice if you could try again.

emkornfield replied:
done, seems to work.

emkornfield commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
+      // when the leaf-values are nullable (otherwise no spacing is needed),
+      // Because only nested lists exists it is sufficient to know that the field
+      // was either null or included it (i.e. >= previous definition level -> > previous
+      // definition - 1). If there where structs mixed in, we need to know the def_level
```
fix comment should "- 2"

emkornfield replied:
I think "- 1" might be confusing here because we subtract below.

emkornfield commented:
CC @wesm @pitrou I think this is ready for review now.

pitrou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -40,12 +40,13 @@ if(ARROW_CPU_FLAG STREQUAL "x86")
     check_cxx_compiler_flag(${ARROW_SSE4_2_FLAG} CXX_SUPPORTS_SSE4_2)
   endif()
   check_cxx_compiler_flag(${ARROW_AVX2_FLAG} CXX_SUPPORTS_AVX2)
+  check_cxx_compiler_flag(${ARROW_AVX2_FLAG} CXX_SUPPORTS_AVX2)
```
This doesn't seem useful :-)

pitrou commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
 
+namespace {
+
+inline void CheckLevelRange(const int16_t* levels, int64_t num_levels,
```
I think it would be nice to move the low-level handling of levels to a dedicated C++ file.

pitrou commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
+#endif  // ARROW_HAVE_BMI2
+
+  } else {
+    // No Special intsturction are used for non-repeated case.
```
"instructions"

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+  return mask;
+}
+
+/// Builds a  bitmap where each set bit indicates the correspond level is greater
```
"corresponding"?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+
+/// Builds a  bitmap where each set bit indicates the correspond level is greater
+/// than rhs.
+static inline int64_t GreaterThanBitmap(const int16_t* levels, int64_t num_levels,
```
Why does it return `int64_t` rather than `uint64_t`?

kiszk replied:
Is this function used only for a little-endian platform?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+/// \param[in] valid_bits_offset The bit-offset at which to start appending new bits.
+/// \param[in,out] valid_bits The validity bitmap to append to.
+/// \returns The new bit offset inside of valid_bits.
+static inline int64_t AppendBitmap(uint64_t new_bits, int64_t number_of_bits,
```
Function name is far too generic. How about `AppendWordToBitmap`?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+
+/// Append bits number_of_bits from new_bits to valid_bits and valid_bits_offset.
+///
+/// \param[in] new_bits The zero-padded bitmap to append.
```
"zero-padded" doesn't give a clue. Are the `number_of_bits` LSBs appended?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+    // big-endian/MSB and shifts right move the bits to the left (division).
+    // 0  1  2  3  4  5  6  7   8  9  10 11 12 13 14 15   16 17 18 19 20 21 22 23
+    // Shifted mask should look like this assuming bit offset = 6:
+    // 2  3  4  5  6  7  N  N   10 11 12 13 14 15  N  N   18 19 20 21 22 23  N  N
```
??? Why are the "N"s lost here?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+                                   int64_t valid_bits_length, int64_t valid_bits_offset,
+                                   uint8_t* valid_bits) {
+  // Selection masks to retrieve all low order bits for each bytes.
+  constexpr uint64_t kLsbSelectionMasks[] = {
```
This entire function looks cryptic.
1) ideally this would be in `bit-util.h`
2) why not add something like a `AppendWord(uint64_t word, int64_t nbits)` to `BitmapWriter` or `FirstTimeBitmapWriter`?
3) the implementation looks gratuitously complicated

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+}
+
+/// \brief Appends bit values to the validitdy bimap_valid bits, based on bitmaps
+/// generated by GreaterThanBitmap, and the appropriate treshold definition_leve.
```
I don't think it makes sense for this to be a helper function.

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+                         validity_bitmap_offset, set_bit_count);
+  return selected_count;
+#else
+  // We shouldn't get here.
```
Then please only define this function if `ARROW_HAVE_BMI2` is defined.

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -0,0 +1,162 @@
+
+using ::testing::ElementsAreArray;
+
+std::string ToString(const uint8_t* bitmap, int64_t bit_count) {
```
Please, a more specific name (such as `BitmapToString`).

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -0,0 +1,162 @@
+  return arrow::internal::Bitmap(bitmap, /*offset*/ 0, /*length=*/bit_count).ToString();
+}
+
+std::string ToString(const std::vector<uint8_t>& bitmap, int64_t bit_count) {
```
Ditto here.

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -0,0 +1,162 @@
+}
+
+TEST(TestAppendSelectedBitsToValidityBitmap, BasicOperation) {
+#if !defined(ARROW_HAVE_BMI2)
```
Then just don't define the test.

kiszk commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
+#else  // big-endian
+    // Optimized SIMD uses bit shifts that are unlikely to work on big endian platforms.
+    DefinitionLevelsToBitmapScalar(def_levels, num_def_levels, max_definition_level,
+                                   max_repitition_level, values_read, null_count,
```
typo: `max_repitition_level` -> `max_repetition_level`

kiszk commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
+    DefinitionLevelsToBitmapSimd</*has_repeated_parent=*/false>(
+        def_levels, num_def_levels, max_definition_level, values_read, null_count,
+        valid_bits, valid_bits_offset);
+  }
+}
```
Do we need to move these two lines before line 186?

kiszk commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -50,6 +51,140 @@ using arrow::internal::checked_cast;
+    uint64_t defined_bitmap = internal::GreaterThanBitmap(def_levels, batch_size,
+                                                          required_definition_level - 1);
+    if (has_repeated_parent) {
+      // This is currently a specialized code path assuming only (nested) lists
```
Since this path is executed only when #if defined(ARROW_HAVE_BMI2) is true, it would be good to add such as code

```
#if defined(ARROW_HAVE_BMI2)
...
#else
  assert(false && "must not execute this without BMI2");
#endif
```

kiszk commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -0,0 +1,162 @@
+  return ToString(bitmap.data(), bit_count);
+}
+
+TEST(TestGreaterThanBitmap, GeneratesExpectedBitmasks) {
```
Most of these tests fail on big-endian platform

```
[  FAILED  ] TestAppendBitmap.TestOffsetOverwritesCorrectBitsOnExistingByte
[  FAILED  ] TestAppendBitmap.TestOffsetShiftBitsCorrectly
[  FAILED  ] TestAppendBitmap.AllBytesAreWrittenWithEnoughSpace
[  FAILED  ] TestAppendBitmap.OnlyApproriateBytesWrittenWhenLessThen8BytesAvailable
[  FAILED  ] TestAppendToValidityBitmap.BasicOperation
```

emkornfield replied:
There was a substantial refactoring, I tried to provide guards but without a big-endian machine in CI it will be very hard to to catch all of these issues.

kiszk commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,191 @@
+  int64_t bit_offset = valid_bits_offset % 8;
+
+  int64_t new_offset = valid_bits_offset + number_of_bits;
+  union ByteAddressableBitmap {
```
Why do we need to use union? `bytes[]` is used only at line 106.   
It looks simple to explicitly extract the lowest 8bit instead of using union.

emkornfield commented:
@pitrou I think I addressed your comments.  One of them that went stale was the complexity for "AppendWord", I tried to remove parts that did not seem to affect performance on benchmarks for parquet column reading but I spent some time trying to maximize word level parallelism for the unaligned case, because I think at least for repeated fields I expect this case to be fairly common. 

As a point of comparison using the AppendWord implementation I put in place for BigEndian shows much smaller improvements.  Really this should use CopyBitmap but I didn't want to start moving a move code then I needed to in bit_util.h for this PR. I opened 
 a [JIRA (ARROW-8595)](https://issues.apache.org/jira/browse/ARROW-8595) to track some improvements in that regard.

If more comments are needed or you think there is a cleaner way of writing the code, I'm happy for input.

@wesm do you want to look at the parquet specific logic/comments to make sure I captured them correctly?

wesm commented:
I'll try to have a closer look tomorrow or Tuesday

emkornfield commented:
@wesm wanted to make sure this is still on your radar

wesm commented:
Yep sorry thanks for the nudge, will look today

wesm commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,101 @@ class FirstTimeBitmapWriter {
+      // N  N  Y  Y  Y  Y  Y  Y   N  N   Y  Y  Y  Y  Y  Y    N  N  Y  Y  Y  Y  Y  Y
+      // So shifted_word should look like;
+      // 2  3  4  5  6  7  N  N   10 11 12 13 14 15  N  N   18 19 20 21 22 23  N  N
+      // clang-format on
```
remove this?

wesm commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,101 @@ class FirstTimeBitmapWriter {
+    // So the new state is either the last relevant byte in 'word'
+    // or zero if we happen to be at a fresh byte.
+    current_byte_ =
+        bit_mask_ != 0x1 ? *(reinterpret_cast<uint8_t*>(&word) + bytes_for_word - 1) : 0;
```
This section is a bit too much bit manipulation for me to carefully scrutinize but I trust you. There are some typos in the comments

emkornfield replied:
OK, if there is anything I can add for documentation please let me know.

pitrou replied:
This is cryptic. Can you rewrite it in a more understandable way? For example (untested):
```cpp
if (bit_mask_ == 0x1) {
  current_byte_ = 0;
} else {
  current_byte_ = *(append_position + bytes_for_word - 1);
}
```

wesm commented:
I think this is fine to merge once most of the typos in the comments are fixed. A rebase will probably fix the Rust lint error

emkornfield commented:
> One problem with introducing more SIMD code is that we won't yet have a runtime dispatching strategy. We will need to go through all of our SIMD accelerations in this library and refactor things so that we can build a "fat" binary that includes both AVX2/BMI2-accelerated versions and the non-SIMD versions. That way we can reap the benefits of this work in portable packages like Python wheels / conda packages.

I did a little research and I think I might lean towards macros around [FMV](https://gcc.gnu.org/wiki/FunctionMultiVersioning) for clang/GCC that could enable fallback to a slow version for MSVC (or if we are ambitious maybe generating IF/DEFs accordingly.  I can't seem to find the JIRA that covered this.

wesm commented:
What I've seen other projects do (have to dig for some examples) is to have files like

```
functionality_nosimd.cc
functionality_sse42.cc
functionality_avx2.cc
```

The appropriate compiler flags are set on the respective compilation unit. 

Then query the CPU at runtime to determine its features and choose which variant to invoke. @cyb70289 expressed an interest in this recently, I think

wesm commented:
Take a look at how this is currently being handled in NumPy

* https://numpy.org/neps/nep-0038-SIMD-optimizations.html
* https://github.com/numpy/numpy/pull/13516

I assume a lot of thought was put into this, and since it was done in 2019 it should provide a good guideline for what we should do

wesm commented:
The 32-bit R failure seems like it could be real

emkornfield commented:
@nealrichardson i'm not sure what the error is saying?  also 32-bit R didn't realize that is still a thing :)

cyb70289 commented:
> Take a look at how this is currently being handled in NumPy
> 
> * https://numpy.org/neps/nep-0038-SIMD-optimizations.html
> * [numpy/numpy#13516](https://github.com/numpy/numpy/pull/13516)
> 
> I assume a lot of thought was put into this, and since it was done in 2019 it should provide a good guideline for what we should do

This is very useful resource to learn, thanks @wesm. We definitely need a better simd code infrastructure if more simd optimization is to be added.

nealrichardson commented:
@emkornfield yeah everything is run twice, 32-bit and 64-bit, because Windows. 

Usually when there's an failure with no clear error, it means that the process segfaulted. Based on the code that's running (https://github.com/apache/arrow/blob/master/r/R/parquet.R#L442-L449) and the output we do see before the crash, my guess is that ReadTable is crashing. What's odd is that that method is called, even with the same test file, several times in r/tests/testthat/test-parquet.R, and those don't crash.

One difference I notice is in how the file connection is opened; not sure how your patch would make that suddenly matter though. But you could try applying this diff and see if the failure goes away:

```
diff --git a/r/R/parquet.R b/r/R/parquet.R
index 8d363e4cb..1415a06da 100644
--- a/r/R/parquet.R
+++ b/r/R/parquet.R
@@ -439,7 +439,7 @@ ParquetFileWriter$create <- function(
 #' @export
 #' @examples
 #' \donttest{
-#' f <- system.file("v0.7.1.parquet", package="arrow")
+#' f <- ReadableFile$create(system.file("v0.7.1.parquet", package="arrow"))
 #' pq <- ParquetFileReader$create(f)
 #' pq$GetSchema()
 #' if (codec_is_available("snappy")) {
@@ -447,6 +447,7 @@ ParquetFileWriter$create <- function(
 #'   tab <- pq$ReadTable(starts_with("c"))
 #'   tab$schema
 #' }
+#' f$close()
 #' }
 #' @include arrow-package.R
 ParquetFileReader <- R6Class("ParquetFileReader",
diff --git a/r/man/ParquetFileReader.Rd b/r/man/ParquetFileReader.Rd
index 1e9f78f4d..3e052ae20 100644
--- a/r/man/ParquetFileReader.Rd
+++ b/r/man/ParquetFileReader.Rd
@@ -33,7 +33,7 @@ with columns filtered by a character vector of column names or a
 
 \examples{
 \donttest{
-f <- system.file("v0.7.1.parquet", package="arrow")
+f <- ReadableFile$create(system.file("v0.7.1.parquet", package="arrow"))
 pq <- ParquetFileReader$create(f)
 pq$GetSchema()
 if (codec_is_available("snappy")) {
@@ -41,5 +41,6 @@ if (codec_is_available("snappy")) {
   tab <- pq$ReadTable(starts_with("c"))
   tab$schema
 }
+f$close()
 }
 }
```

wesm commented:
> I might lean towards macros around FMV for clang/GCC that could enable fallback to a slow version for MSVC

FTR, it would seem unfortunate to do the work of SIMD-ifying code and then have it not be availabe on Windows (e.g. Python and R have a lot of Windows users). I'm not sure how much additional work is needed to get things to work on MSVC but it would be be good to have an OS-independent approach

nealrichardson commented:
For what it's worth, R on Windows uses mingw, not msvc

emkornfield commented:
The separate .cc files might work as, we should probably cleanup macro naming so we can distinguish between min and max SSE instruction se to use (I imagine some people won't want code bloat).

emkornfield commented:
@nealrichardson looks like a different error now?  I think I probably want to use the scalar version for 32-bit builds.  Is there an example in the code that provide the right compiler checks to do this?

nealrichardson commented:
@emkornfield that looks like the same R Windows 32-bit failure to me. 

I'm not sure I understand your other question. Are you saying you want to use the old (status quo) code on 32-bit? That sounds like something you'd detect in cmake to me. If the C++ library needs to be built with different flags, the R Windows build script is in ci/scripts/PKGBUILD.

wesm commented:
According to https://blog.kowalczyk.info/article/j/guide-to-predefined-macros-in-c-compilers-gcc-clang-msvc-etc..html the define is `__MINGW32__`. I checked an `-m32` build on Linux and it runs fine

pitrou commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,103 @@ class FirstTimeBitmapWriter {
+      uint64_t captured_carry = carry_mask & word;
+      // mask_carry_bits is:
+      // N  N  N  N  N  N  8  9   N  N  N  N  N   N 16 17    N  N   N  N  N  N  N  N
+      uint64_t mask_carry_bits = (captured_carry >> 8) << bit_offset;
```
This is exactly `(captured_carry >> bits_to_carry)`, right? So in the end we could just write `word = word >> bits_to_carry`?
(which means most of the above could be simplified away)

emkornfield replied:
yes, I confused myself initially on LSB/little-endian and never recovered from that.  removed a bunch of code from this.  Thank you.

pitrou commented on cpp/src/arrow/util/bit_util_test.cc:
```diff
@@ -315,6 +318,115 @@ TEST(FirstTimeBitmapWriter, NormalOperation) {
+
+TEST(FirstTimeBitmapWriter, AppendWordOffsetOverwritesCorrectBitsOnExistingByte) {
+  auto check_append = [](const std::string& expected_bits, int64_t offset) {
+    std::vector<uint8_t> valid_bits = {0x00};
```
For each of the tests here it would be nice to have two variants or iterations: one that starts with a bitmap full of 0s, and one with a bitmap full of 1s.

emkornfield replied:
I added variants to this test.  Note that I don't think have a bitmap full of 111s iand overwriting trailing ones isn't really expected because any not processed are expected to be zero.

pitrou replied:
Well, `FirstTimeBitmapWriter` is supposed to work on uninitialized data.

pitrou commented on cpp/src/arrow/util/bit_util_test.cc:
```diff
@@ -315,6 +318,115 @@ TEST(FirstTimeBitmapWriter, NormalOperation) {
+  EXPECT_EQ(valid_bits[0], 0x2);
+}
+
+TEST(FirstTimeBitmapWriter, AppendLessThenByte) {
```
Typo: "LessThan"

emkornfield replied:
fixed.

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -0,0 +1,170 @@
+
+  // TODO(itaiin): As an interim solution we are splitting the code path here
+  // between repeated+flat column reads, and non-repeated+nested reads.
+  // Those paths need to be merged in the future
```
Is this TODO still relevant? Is it a performance issue or a behaviour issue?

emkornfield replied:
Yes it is.  It is a behaviour issue. hopefully i'll be addressing this over the next few PRs.  This was a first step.

emkornfield replied:
To expand on this, this really need to accept a second repetition level which is repeated_parent_definition_level that would replace max_definition_level - 1 below.

pitrou commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,103 @@ class FirstTimeBitmapWriter {
+        bit_mask_ != 0x1 ? *(reinterpret_cast<uint8_t*>(&word) + bytes_for_word - 1) : 0;
+
+#else  // big-endian
+    BitmapReader reader(reinterpret_cast<uint8_t*>(&word), 0, /*length=*/number_of_bits);
```
Uh, will this actually work in big-endian? This is going to read the most significant byte first... I think you need to shift the word logically, like this:
```cpp
for (int x = 0; x < number_of_bits; x++) {
  if (word & 0x1) {
    Set();
  } else {
    Clear();
  }
  Next();
  word >>= 1;
}
```

emkornfield replied:
That does look more correct.  Without big-endian in CI its hard to confirm.

emkornfield commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,71 @@ class FirstTimeBitmapWriter {
+    }
+
+#else  // big-endian
+    for (int x = 0; x < number_of_bits; x++) {
```
maybe I should just remove this since I can't verify?

fsaintjacques replied:
An aborting static_assert such that one does not get a surprise at runtime?

fsaintjacques commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -0,0 +1,170 @@
+                              const int16_t max_repetition_level, int64_t* values_read,
+                              int64_t* null_count, uint8_t* valid_bits,
+                              int64_t valid_bits_offset) {
+  if (max_repetition_level > 0) {
```
The interlacing of 1 runtime branch and 2 static branches is hard to read. It seems the big-endian version doesn't even close the block with a corresponding parenthesis.

emkornfield replied:
split this into two functions.

fsaintjacques commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,67 @@
+/// than rhs.
+static inline uint64_t GreaterThanBitmap(const int16_t* levels, int64_t num_levels,
+                                         int16_t rhs) {
+  return LevelsToBitmap(levels, num_levels, [&](int16_t value) { return value > rhs; });
```
You should just capture rhs by value.

emkornfield replied:
done.

fsaintjacques commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -0,0 +1,67 @@
+  // Both clang and GCC can vectorize this automatically with SSE4/AVX2.
+  uint64_t mask = 0;
+  for (int x = 0; x < num_levels; x++) {
+    mask |= static_cast<int64_t>(predicate(levels[x]) ? 1 : 0) << x;
```
- Mixing `uint64_t` and `int64_t` in the mask update.
- Add a DCHECK_LE(num_levels, 64).

emkornfield replied:
fixed the first point.  I think this header might get exposed publicly at some point, so  we can't use DCHECK I think.

fsaintjacques commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -0,0 +1,170 @@
+                                                  /*length=*/num_def_levels);
+  int64_t set_count = 0;
+  *values_read = 0;
+  while (num_def_levels > 0) {
```
The number of iterations is only dependent on num_def_levels. Break the loop in increments of kBitMaskSize and process the leftover if needed. This will ensure that `batch_size` is statically known to be `kBitMaskSize` inside the loop, and thus the compiler will be able to apply constant propagation and forward to `GreaterThanBitmap`.

fsaintjacques commented:
@emkornfield take a look at the small signature change I added in 0af944d.

nealrichardson commented:
I'm curious, what fixed the crash we observed on 32-bit Windows R?

fsaintjacques commented:
@nealrichardson seems like the R CI integration doesn't run anymore.

nealrichardson commented:
Well, that's one way to fix it.

nealrichardson commented:
Looks like they stopped running in https://github.com/apache/arrow/commit/4116516f22b680b70ab32ea28c029cc47e8d8988#diff-2a0471f8b081b56e67bf646767021982

nealrichardson commented:
IMO we should revert this change: https://github.com/apache/arrow/commit/4116516f22b680b70ab32ea28c029cc47e8d8988#diff-2a0471f8b081b56e67bf646767021982

It seems only to obfuscate, and it seems it might not work.

emkornfield commented:
@pitrou did you want to take another look?

@nealrichardson should I hold off until we an revert the change to see if the test still fails?

pitrou commented:
I see a slight slowdown here:
* git master:
```
-------------------------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      3817689 ns      3814696 ns          733 bytes_per_second=10.24G/s
BM_ReadColumn<true,Int32Type>      27211615 ns     27194320 ns           99 bytes_per_second=2.15463G/s
BM_ReadColumn<false,Int64Type>      7085595 ns      7079927 ns          399 bytes_per_second=11.0347G/s
BM_ReadColumn<true,Int64Type>      29967050 ns     29951421 ns           89 bytes_per_second=3.26049G/s
BM_ReadColumn<false,DoubleType>     7098703 ns      7093371 ns          398 bytes_per_second=11.0138G/s
BM_ReadColumn<true,DoubleType>     30548935 ns     30532825 ns           89 bytes_per_second=3.1984G/s
BM_ReadColumn<false,BooleanType>   11931142 ns     11928150 ns          235 bytes_per_second=838.353M/s
BM_ReadColumn<true,BooleanType>    29983466 ns     29970561 ns           92 bytes_per_second=1000.98M/s
BM_ReadIndividualRowGroups         15547894 ns     15538074 ns          183 bytes_per_second=6.28497G/s
BM_ReadMultipleRowGroups           15692019 ns     15682364 ns          182 bytes_per_second=6.22714G/s
```
* this PR:
```
-------------------------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      3819328 ns      3816352 ns          740 bytes_per_second=10.2356G/s
BM_ReadColumn<true,Int32Type>      30445380 ns     30429672 ns           89 bytes_per_second=1.92555G/s
BM_ReadColumn<false,Int64Type>      7134237 ns      7128926 ns          397 bytes_per_second=10.9589G/s
BM_ReadColumn<true,Int64Type>      35230806 ns     35215688 ns           81 bytes_per_second=2.77309G/s
BM_ReadColumn<false,DoubleType>     7007725 ns      7001947 ns          394 bytes_per_second=11.1576G/s
BM_ReadColumn<true,DoubleType>     33062963 ns     33046208 ns           83 bytes_per_second=2.95514G/s
BM_ReadColumn<false,BooleanType>   12116716 ns     12113217 ns          232 bytes_per_second=825.545M/s
BM_ReadColumn<true,BooleanType>    33176910 ns     33163314 ns           84 bytes_per_second=904.614M/s
BM_ReadIndividualRowGroups         18253632 ns     18243651 ns          160 bytes_per_second=5.35289G/s
BM_ReadMultipleRowGroups           18032877 ns     18022962 ns          157 bytes_per_second=5.41843G/s
```

pitrou commented:
Rebased.

pitrou commented:
I don't understand the Windows GLib build failure. @kou

emkornfield commented:
@pitrou what flags/CPU are you running this under?  I'll recheck on my box later.

pitrou commented:
It's a AMD Ryzen 9 3900X with gcc 7.5, with `-DARROW_SIMD_LEVEL=AVX2`.

kou commented:
(It seems that Windows GLib build failure has gone without any change.)

emkornfield commented:
OK it seems like clang and GCC perform very differently.

Clang-8 new code vs master:
![image](https://user-images.githubusercontent.com/17869838/81888336-ff86bb00-9555-11ea-97da-8becce12bee2.png)

GCC 7.5 (inline with the results from @pitrou):
![image](https://user-images.githubusercontent.com/17869838/81888475-3fe63900-9556-11ea-98d6-a1dff1ea24e2.png)

Interesting it looks like GCC 7.5 does better on the write path (GCC vs Clang on master):
![image](https://user-images.githubusercontent.com/17869838/81888643-a53a2a00-9556-11ea-821f-8439043da4c8.png)

emkornfield commented:
@pitrou Doing some investigation it appears GCC 7.5 fails to auto-vectorize the comparison of levels.  GCC 8.3 does autovectorize.  Are we stuck with GCC 7.5 for releasing?

pitrou commented:
I have no idea :-) I simply use gcc 7.5 locally, but I guess we're not stuck with it. And gcc 10 was released AFAIK, so it's starting to get old as well.

pitrou commented:
Ok, here is with clang 9.0.
* Before:
```
BM_ReadColumn<false,Int32Type>      3907221 ns      3904115 ns          179 bytes_per_second=10.0055G/s
BM_ReadColumn<true,Int32Type>      24519395 ns     24505645 ns           28 bytes_per_second=2.39103G/s
BM_ReadColumn<false,Int64Type>      7180572 ns      7174186 ns           98 bytes_per_second=10.8897G/s
BM_ReadColumn<true,Int64Type>      29619694 ns     29608069 ns           23 bytes_per_second=3.2983G/s
BM_ReadColumn<false,DoubleType>     7441291 ns      7434806 ns           86 bytes_per_second=10.508G/s
BM_ReadColumn<true,DoubleType>     30991035 ns     30975255 ns           24 bytes_per_second=3.15272G/s
BM_ReadColumn<false,BooleanType>   10626337 ns     10622931 ns           66 bytes_per_second=941.36M/s
BM_ReadColumn<true,BooleanType>    29892031 ns     29878906 ns           23 bytes_per_second=1004.05M/s
BM_ReadIndividualRowGroups         15414099 ns     15402186 ns           43 bytes_per_second=6.34042G/s
BM_ReadMultipleRowGroups           15087609 ns     15078640 ns           42 bytes_per_second=6.47646G/s
```
* After:
```
BM_ReadColumn<false,Int32Type>      3858847 ns      3855927 ns          181 bytes_per_second=10.1305G/s
BM_ReadColumn<true,Int32Type>      17606021 ns     17594549 ns           38 bytes_per_second=3.33022G/s
BM_ReadColumn<false,Int64Type>      7164074 ns      7158771 ns           98 bytes_per_second=10.9132G/s
BM_ReadColumn<true,Int64Type>      22593565 ns     22579690 ns           29 bytes_per_second=4.32496G/s
BM_ReadColumn<false,DoubleType>     7399623 ns      7393991 ns           97 bytes_per_second=10.566G/s
BM_ReadColumn<true,DoubleType>     22415125 ns     22400757 ns           27 bytes_per_second=4.35951G/s
BM_ReadColumn<false,BooleanType>   10240053 ns     10236365 ns           68 bytes_per_second=976.909M/s
BM_ReadColumn<true,BooleanType>    22663603 ns     22651544 ns           31 bytes_per_second=1.29337G/s
BM_ReadIndividualRowGroups         11406811 ns     11398414 ns           61 bytes_per_second=8.56753G/s
BM_ReadMultipleRowGroups           11314315 ns     11305522 ns           56 bytes_per_second=8.63792G/s
```

pitrou commented:
As a sidenote, it seems BMI2 may be problematic on AMD CPUs:
https://twitter.com/trav_downs/status/1202793097962364928

emkornfield commented:
> As a sidenote, it seems BMI2 may be problematic on AMD CPUs:
https://twitter.com/trav_downs/status/1202793097962364928

Yea we need some benchmarks to see if the performance is worth it for AMD (I suspect it might make this patch break-even instead of an improvement) but don't have hardware to test.

pitrou commented:
I do have an AMD CPU, so I'll gladly test if you add some benchmarks :-)

emkornfield commented:
@pitrou https://github.com/apache/arrow/pull/6985/commits/a36dc55efe3def27eef413b8e17567ec400a01ff adds a benchmark.  Should be cherry-pickable to a clean branch for comparison.

On my machine it shows ~2X speedup.  

```
Before:
---------------------------------------------------------------------------------------------------------                                      
Benchmark                                               Time             CPU   Iterations UserCounters...                                      
---------------------------------------------------------------------------------------------------------                                      
BM_DefinitionLevelsToBitmapRepeatedAllMissing        2314 ns         2314 ns       302395 bytes_per_second=844.019M/s                          
BM_DefinitionLevelsToBitmapRepeatedAllPresent        2333 ns         2333 ns       301106 bytes_per_second=837.088M/s                          
BM_DefinitionLevelsToBitmapRepeatedMostPresent       2458 ns         2458 ns       286182 bytes_per_second=794.611M/s   
```


```
After:
---------------------------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------                                      
BM_DefinitionLevelsToBitmapRepeatedAllMissing        1054 ns         1054 ns       665201 bytes_per_second=1.80969G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1218 ns         1218 ns       577196 bytes_per_second=1.56564G/s                          
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1288 ns         1288 ns       542522 bytes_per_second=1.48129G/s 

```

pitrou commented:
Here are some benchmarks on my machine, with gcc 7.5:
* ARROW_SIMD_LEVEL=NONE:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         965 ns          965 ns      2895940 bytes_per_second=1.97577G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1565 ns         1565 ns      1835474 bytes_per_second=1.21861G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1709 ns         1709 ns      1628131 bytes_per_second=1.11626G/s
```
* ARROW_SIMD_LEVEL=AVX2:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing        3726 ns         3726 ns       748789 bytes_per_second=524.229M/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        5781 ns         5780 ns       484016 bytes_per_second=337.895M/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       5776 ns         5775 ns       485316 bytes_per_second=338.183M/s
```

With clang 8.0:
* ARROW_SIMD_LEVEL=NONE:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         935 ns          935 ns      3009389 bytes_per_second=2.04013G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1326 ns         1326 ns      2152178 bytes_per_second=1.43859G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1601 ns         1601 ns      1789999 bytes_per_second=1.19129G/s
```
* ARROW_SIMD_LEVEL=AVX2:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing        1126 ns         1126 ns      2483214 bytes_per_second=1.69443G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        3044 ns         3043 ns       909419 bytes_per_second=641.808M/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       2873 ns         2872 ns       962839 bytes_per_second=679.975M/s
```

So the BMI2 instructions incur a large slowdown here (even catastrophic with gcc).

emkornfield commented:
> So the BMI2 instructions incur a large slowdown here (even catastrophic with gcc).

I would expect bad numbers with GCC 7.5 since we lose autovectorization and use BMI2.  Agree that clang-8 numbers still look pretty bad on AMD.  I guess this PR will have to wait until we can do runtime dispatch :(

emkornfield commented:
@pitrou I updated this PR to require AVX512 for the use of BMI2 codepath.  I think this should prevent issues for AMD architectures (I don't believe there are any that support AVX512) until we can do runtime dispatch for this issue.

wesm commented:
I'll kick the tires on this re: performance. Is this ready to merge otherwise?

emkornfield commented:
As long as @pitrou is happy with my workaround to avoid inflicting pain on AMD processor users.

I think I need to fix one signed conversion bug to make CI happy.  I should be able to get to that tonight.

emkornfield commented:
OK build is green (side note I don't think this will compile on s390x, not sure if that is a blocker).

pitrou commented:
Here are some benchmarks on my machine, with gcc 7.5:
* `ARROW_SIMD_LEVEL=AVX2`:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         934 ns          934 ns      3004804 bytes_per_second=2.04274G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1327 ns         1327 ns      1908147 bytes_per_second=1.4377G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1725 ns         1725 ns      1649108 bytes_per_second=1.10569G/s
```
* `ARROW_SIMD_LEVEL=SSE4_2`:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing        1384 ns         1384 ns      2029778 bytes_per_second=1.37806G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        2054 ns         2053 ns      1247469 bytes_per_second=951.163M/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       2124 ns         2124 ns      1303578 bytes_per_second=919.503M/s
```
* `ARROW_SIMD_LEVEL=NONE`:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         925 ns          925 ns      3025393 bytes_per_second=2.06245G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1505 ns         1504 ns      1881938 bytes_per_second=1.26785G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1725 ns         1725 ns      1598820 bytes_per_second=1.10599G/s
```

So it seems that gcc's SSE4.2 auto-vectorization may lead to suboptimal code, but that's not a problem for this PR.

pitrou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -42,7 +42,7 @@ if(ARROW_CPU_FLAG STREQUAL "x86")
   else()
     set(ARROW_SSE4_2_FLAG "-msse4.2")
-    set(ARROW_AVX2_FLAG "-mavx2")
+    set(ARROW_AVX2_FLAG "-mavx2 -mbmi2")
```
Shouldn't you put this under AVX512?

pitrou commented on cpp/src/arrow/util/bit_util.h:
```diff
@@ -610,6 +618,61 @@ class FirstTimeBitmapWriter {
+  ///            to be unset (i.e. 0).
+  /// \param[in] number_of_bits The number of bits to append from word.
+  void AppendWord(uint64_t word, int64_t number_of_bits) {
+#if defined(ARROW_LITTLE_ENDIAN)
```
Should be `#if ARROW_LITTLE_ENDIAN`.

emkornfield commented:
@pitrou I think the C GLib build failure looks unrelated to me.  I fixed the  s390x build.

emkornfield commented:
@wesm I expect your PR to get merged first but if it doesn't consider using PopCount methods in this one.

wesm commented:
+1. I'm going to go ahead and merge this and then I'll rebase #7352
